smk/include/bwa_index_wrapper.smk

```diff
@@ -0,0 +1,226 @@
+            "{somewhere}/BWA_index/{something}.{fasta}.amb",
+            "{somewhere}/BWA_index/{something}.{fasta}.ann",
+            "{somewhere}/BWA_index/{something}.{fasta}.bwt.2bit.64",
+            "{somewhere}/BWA_index/{something}.{fasta}.pac",
+        log:
+            "{somewhere}/BWA_index/BWA_index.{something}.{fasta}.log"
+        wildcard_constraints:
+            fasta='fasta|fna'
+        resources:
+            mem = lambda wildcards, input, attempt: 5 + int(22*input.size_mb/1024) + 40*(attempt-1),
+        threads: 4
+        conda:
+            config["minto_dir"]+"/envs/MIntO_base.yml" #bwa-mem2
+        shell:
+            """
+            outdir=$(dirname {output[0]})
+            umask 002 && mkdir -p {CLUSTER_LOCAL_DIR}/$outdir
+            cd {CLUSTER_LOCAL_DIR}/$outdir
+            prefix=$(basename {input})
+            time (
+                bwa-mem2 index {input.fasta} -p $prefix
+                echo "NODE: $(hostname)"
+                echo "SENDING:"
+                echo "$(pwd)/$prefix --> $outdir"
+                rsync -av --itemize-changes $prefix.* $outdir/
+            ) >& {log}
+            """
+
+    checkpoint prepare_files_for_syncing:
+        input:
+            "{somewhere}/BWA_index/{something}.{fasta}.0123",
+            "{somewhere}/BWA_index/{something}.{fasta}.amb",
+            "{somewhere}/BWA_index/{something}.{fasta}.ann",
+            "{somewhere}/BWA_index/{something}.{fasta}.bwt.2bit.64",
+            "{somewhere}/BWA_index/{something}.{fasta}.pac",
+        output:
+            directory("{somewhere}/BWA_index/{something}.{fasta}.sync"),
+        log:
+            "{somewhere}/BWA_index/sync.{something}.{fasta}.log"
+        wildcard_constraints:
+            fasta='fasta|fna'
+        localrule: True
+        shell:
+            """
+            dir="{output}"
+            mkdir -p $dir
+            for node in {CLUSTER_NODES}; do
+                touch $dir/$node
+            done
+            """
+
+    def get_node_request_argument(node, cluster_workload_manager):
+        if cluster_workload_manager.upper() == 'SLURM':
+            return f"--nodelist {node}"
+        elif cluster_workload_manager.upper() in['PBS', 'TORQUE']:
+            return f"-l nodes={node}"
+        elif cluster_workload_manager.upper() == 'SGE':
+            return f"-l h={node}"
+        elif cluster_workload_manager.upper() == 'LSF':
+            return f"-m {node}"
+        else:
+            raise Exception(f"MIntO error: cannot handle cluster workload manager '{cluster_workload_manager}'")
+
+    rule rsync_index_files_to_local_dir:
+        input:
+            "{somefasta}.sync/{node}"
+        output:
+            "{somefasta}.sync/{node}.done"
+        log:
+            "{somefasta}.sync/{node}.log"
+        resources:
+            mem = 2,
+            qsub_args = lambda wildcards: get_node_request_argument(wildcards.node, CLUSTER_WORKLOAD_MANAGER)
+        wildcard_constraints:
```

Comment:
Is `resources.qsub_args` a snakemake variable that gets passed to the generic cluster function, or is there a specific way to include it when running snakemake? I couldn't find it via google.

Reply:
Good catch. It is not automatic. It needs to be passed via `--cluster`. We can define anything within `resources` in a rule and use it when submitting to cluster. I added these details in the `smk` file for now. But we should clarify in the documentation/tutorial. One needs to run snakemake with these arguments:
```
--default-resources gpu=0 mem=4 "qsub_args=''" --cluster 'sbatch -J {name} --mem={resources.mem}G --gres=gpu:{resources.gpu} -c {threads} {resources.qsub_args}'
```

Reply:
Explanation added: b57cbd7